### PR TITLE
Add an interpreter for GEM

### PR DESCRIPTION
**Pull request by wence-**
Takes a GEM tree with some optional data bindings for Variable and
VariableIndex nodes and evaluates it.

Lightly tested!

### CODE REVIEW

**Comment by wence-** [comment]
I haven't thought hard about where to allow hooking this in to the compilation pipeline.  It works on the tree that comes out of fem.process either before or after simplification passes.

**Comment by miklos1** [comment]
We could hook it up with the PyOP2 interpreter.

File: tsfc/interpreter.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    idx = val.fids.index(e.index)
+    return Result(val.arr.sum(axis=idx),
+                  val.fids[:idx] + val.fids[idx+1:])
+
+
+@_evaluate.register(gem.ListTensor)  # noqa: not actually redefinition
+def _(e, self):
+    """List tensors just turn into arrays."""
+    ops = [self(o) for o in e.children]
+    assert all(ops[0].fids == o.fids for o in ops)
```
Comment (miklos1) [inline tsfc/interpreter.py]: This is true for ListTensors where were translated from UFL. GEM doesn't have this restriction though.

Comment (miklos1) [inline tsfc/interpreter.py]: Or is it true? It may not be when the FormSplitter is involved.

Comment (wence-) [inline tsfc/interpreter.py]: I can fix this by relaxing the assertion.

Comment (wence-) [inline tsfc/interpreter.py]: But that requires me to think too hard in my hazy state.

**Pull request merged**